File: 1111-minimum-score-triangulation-of-polygon/1111-minimum-score-triangulation-of-polygon.cpp
class Solution {
public:
    int minScoreTriangulation(vector<int>& values) {
        vector<vector<int>> dp(values.size(), vector<int>(values.size(), -1));
        return solve(values, dp, 0, (int)values.size() - 1);
    }
private:
    int solve(vector<int>& values, vector<vector<int>>& dp, int i, int j)
    {
        if(j - i < 2)
        {
            return 0;
        }
        else if(j - i == 2)
        {
            return values[i] * values[i + 1] * values[j];
        }
        if(dp[i][j] == -1)
        {
            dp[i][j] = INT_MAX;
            for(int k = i + 1; k < j; k++)
            {
                dp[i][j] = min(dp[i][j], values[i] * values[k] * values[j] + solve(values, dp, i, k) + solve(values, dp, k, j));
            }
        }
        return dp[i][j];
    }
};

File: 1111-minimum-score-triangulation-of-polygon/1111-minimum-score-triangulation-of-polygon.py
class Solution:
    def solve(self, values: List[int], dp: List[List[int]], i: int, j: int) -> int:
        if j - i < 2:
            return 0
        elif j - i == 2:
            return values[i] * values[i + 1] * values[j]
        if dp[i][j] == -1:
            dp[i][j] = float('inf')
            for k in range(i + 1, j):
                sub = values[i] * values[k] * values[j] + self.solve(values, dp, i, k) + self.solve(values, dp, k, j)
                dp[i][j] = min(dp[i][j], sub)
        return dp[i][j]
    def minScoreTriangulation(self, values: List[int]) -> int:
        dp = [[-1 for j in range(len(values))] for i in range(len(values))]
        return self.solve(values, dp, 0, len(values) - 1)
        